biz.aQute.repository/src/aQute/bnd/repository/maven/provider/IndexFile.java

```diff
@@ -366,14 +375,31 @@ private boolean isMulti(String name) {
 		return result;
 	}
 
+	/**
+	 * This method actually runs NOT in the background, it syncs
+	 *
+	 * @param refreshAction to call after it has finished
+	 * @return
+	 * @throws Exception
+	 */
 	boolean refresh(Runnable refreshAction) throws Exception {
+		Promise<Boolean> result;
 		if (indexFile.lastModified() != lastModified && last + 10000 < System.currentTimeMillis()) {
 			last = System.currentTimeMillis();
-			Promise<Boolean> serializer = serialize(this::load).onResolve(refreshAction);
-			sync(serializer);
-			return true;
+			result = serialize(this::load);
+		} else {
+			Set<Archive> snapshots = archives.keySet()
+				.stream()
+				.filter(Archive::isSnapshot)
+				.collect(Collectors.toSet());
+			if (!snapshots.isEmpty()) {
+				result = serialize(() -> update(snapshots));
+			} else
+				return false;
 		}
-		return false;
+		Promise<Boolean> serializer = result.onResolve(refreshAction);
```

Comment:
Since you use `onResolve`, `serializer` is the same as `result` which means the refresh action will not happen before `sync(serializer)` returns. That is, `onResolve` is not chained. Just making sure this is your intention.

Reply:
The promise was already serialized so this is ok. I've removed the serializer variable to make this more clear.